xpra/server/auth/keycloak_auth.py

```diff
@@ -0,0 +1,113 @@
+        self.client_secret_key = kwargs.pop("client_secret_key", KEYCLOAK_CLIENT_SECRET_KEY)
+        self.redirect_uri = kwargs.pop("redirect_uri", KEYCLOAK_REDIRECT_URI)
+        self.scope = kwargs.pop("scope", KEYCLOAK_SCOPE)
+        self.grant_type = kwargs.pop("grant_type", KEYCLOAK_GRANT_TYPE)
+        kwargs["prompt"] = kwargs.pop("prompt", "keycloak")
+
+        if KEYCLOAK_GRANT_TYPE == "authorization_code":
+          super().__init__(**kwargs)
+          log("keycloak auth: server_url=%s, client_id=%s, realm_name=%s, redirect_uri=%s, scope=%s, grant_type=%s",
+              self.server_url, self.client_id, self.realm_name, self.redirect_uri, self.scope, self.grant_type)
+
+          # Get authorization code
+          client = WebApplicationClient(KEYCLOAK_CLIENT_ID)
+          authorization_url = KEYCLOAK_SERVER_URL + 'realms/' + KEYCLOAK_REALM_NAME + '/protocol/openid-connect/auth'
+          self.salt = client.prepare_request_uri(
+                      authorization_url,
+                      redirect_uri = KEYCLOAK_REDIRECT_URI,
+                      scope = [KEYCLOAK_SCOPE],
+          )
+        else:
+          raise(NotImplementedError("Warning: only grant type \"authorization_code\" is currently supported."))
+
+    def __repr__(self):
+        return "keycloak"
+
+    def get_challenge(self, digests):
+        assert not self.challenge_sent
+        if "keycloak" not in digests:
+            log.error("Error: client does not support keycloak authentication")
+            return None
+        self.challenge_sent = True
+        return self.salt, "keycloak"
+
+    def check(self, response_json) -> bool:
+        assert self.challenge_sent
+        
+        if response_json is None or response_json == "":
+          log.error("keycloak authentication failed: invalid response received from authorization endpoint.")
+          return False
+        
+        #log("response_json: %s", repr(response_json))
+        response = json.loads(response_json)
+
+        if type(response) != dict or ("code" not in response and "error" not in response):
+          log.error("keycloak authentication failed: invalid response received from authorization endpoint.")
+          return False
+
+        log("check(%s)", repr(response))
+
+        if "error" in response:
+          log.error("keycloak authentication failed with error %s: %s", response["error"], response["error_description"])
+          return False
+
+        if "code" in response:
+          # Configure client
+          keycloak_openid = KeycloakOpenID(server_url=KEYCLOAK_SERVER_URL,
+                            client_id=KEYCLOAK_CLIENT_ID,
+                            realm_name=KEYCLOAK_REALM_NAME,
+                            client_secret_key=KEYCLOAK_CLIENT_SECRET_KEY)
+
+          try:
+            # Get well_known
+            #config_well_know = keycloak_openid.well_know()
+            #log("well_known: %s", repr(config_well_know))
```

Comment:
Is there any harm in logging these details?
If not, then uncomment that line.
Don't use `%s` with `repr`, just use `%r`.

Reply:
There is no harm in more logging this, but this does add a useless request making the authentication slower. I uncommented, feel free to comment again.